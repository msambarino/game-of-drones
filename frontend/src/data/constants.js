export const DEFAULT_RULES = {
  moves: [
    { move: 'rock', kills: 'scissors' },
    { move: 'paper', kills: 'rock' },
    { move: 'scissors', kills: 'paper' }
  ],
  roundWinsRequired: 3
}

export const ROUND_RESULTS = {
  DRAW: 'DRAW',
  PLAYER_1_WINS: 'PLAYER_1_WINS',
  PLAYER_2_WINS: 'PLAYER_2_WINS'
}
