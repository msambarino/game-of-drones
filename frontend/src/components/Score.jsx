import React from 'react'
import PropTypes from 'prop-types'
import { ROUND_RESULTS } from 'data/constants'

const formatRoundWinner = (player1, player2, result) => {
  switch (result) {
    case ROUND_RESULTS.DRAW:
      return '--'
    case ROUND_RESULTS.PLAYER_1_WINS:
      return player1
    case ROUND_RESULTS.PLAYER_2_WINS:
      return player2
    default:
      return null
  }
}

const Score = ({ rounds, player1, player2 }) => (
  <table className="table is-hoverable">
    <thead>
      <tr>
        <th>Round</th>
        <th>{`${player1}'s move`}</th>
        <th>{`${player2}'s move`}</th>
        <th>Winner</th>
      </tr>
    </thead>
    <tbody>
      {rounds.map((round, index) => (
        <tr key={index}>
          <td>{index + 1}</td>
          <td>{round.player1Move}</td>
          <td>{round.player2Move}</td>
          <td>{formatRoundWinner(player1, player2, round.result)}</td>
        </tr>
      ))}
    </tbody>
  </table>
)

Score.propTypes = {
  rounds: PropTypes.array.isRequired,
  player1: PropTypes.string.isRequired,
  player2: PropTypes.string.isRequired
}

export default Score
