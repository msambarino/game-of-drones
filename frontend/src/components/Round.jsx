import React from 'react'
import PropTypes from 'prop-types'
import Score from './Score'

class Round extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      isPlayer1Turn: true,
      player1Move: this.props.moves[0].move,
      player2Move: this.props.moves[0].move
    }

    this.handleMoveChange = this.handleMoveChange.bind(this)
    this.handleOkClick = this.handleOkClick.bind(this)
    this._reset = this._reset.bind(this)
  }

  handleMoveChange(event) {
    if (this.state.isPlayer1Turn) {
      this.setState({ player1Move: event.target.value })
    } else {
      this.setState({ player2Move: event.target.value })
    }
  }

  handleOkClick() {
    if (this.state.isPlayer1Turn) {
      this.setState({ isPlayer1Turn: false })
    } else {
      this._reset()
      this.props.handlePlayRound(this.state.player1Move, this.state.player2Move)
    }
  }

  _reset() {
    this.setState({
      isPlayer1Turn: true,
      player1Move: this.props.moves[0].move,
      player2Move: this.props.moves[0].move
    })
  }

  render() {
    const { rounds, moves, player1, player2 } = this.props
    const { isPlayer1Turn, player1Move, player2Move } = this.state

    const player = isPlayer1Turn ? player1 : player2
    const selectedMove = isPlayer1Turn ? player1Move : player2Move
    const round = rounds.length + 1

    return (
      <div className="container">
        <section className="section is-small columns">
          <div className="column">
            <h1 className="title">{`Round ${round}`}</h1>
            <h3 className="subtitle">{`${player}'s Turn`}</h3>
            <div className="field is-horizontal">
              <label className="label field-label is-normal">
                Select Move:
              </label>
              <div className="field-body">
                <div className="field is-narrow">
                  <div className="control">
                    <div className="select is-fullwidth">
                      <select
                        onChange={this.handleMoveChange}
                        value={selectedMove}
                      >
                        {moves.map(move => (
                          <option key={move.move} value={move.move}>
                            {move.move}
                          </option>
                        ))}
                      </select>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div className="control">
              <button
                className="button is-primary"
                onClick={this.handleOkClick}
              >
                Ok
              </button>
            </div>
          </div>
          <div className="column">
            <h3 className="subtitle">Score</h3>
            <Score player1={player1} player2={player2} rounds={rounds} />
          </div>
        </section>
      </div>
    )
  }
}

Round.propTypes = {
  player1: PropTypes.string.isRequired,
  player2: PropTypes.string.isRequired,
  moves: PropTypes.array.isRequired,
  rounds: PropTypes.array.isRequired,
  handlePlayRound: PropTypes.func.isRequired
}

export default Round
