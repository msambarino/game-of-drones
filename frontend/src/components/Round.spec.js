import React from 'react'
import { shallow, mount, render } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import { DEFAULT_RULES } from '../data/constants'
import Round from './Round'
import Score from './Score'

test('snapshot test', () => {
  const component = shallow(
    <Round
      player1="p1"
      player2="p2"
      rounds={[]}
      moves={DEFAULT_RULES.moves}
      handlePlayRound={jest.fn()}
    />
  )
  const tree = shallowToJson(component)
  expect(tree).toMatchSnapshot()
})

test('should render a Score component', () => {
  const component = shallow(
    <Round
      player1="p1"
      player2="p2"
      rounds={[]}
      moves={DEFAULT_RULES.moves}
      handlePlayRound={jest.fn()}
    />
  )

  expect(component.find(Score)).toHaveLength(1)
})

test("should be player2's turn after player1 selects his move", () => {
  const component = shallow(
    <Round
      player1="p1"
      player2="p2"
      rounds={[]}
      moves={DEFAULT_RULES.moves}
      handlePlayRound={jest.fn()}
    />
  )
  const instance = component.instance()

  expect(instance.state.isPlayer1Turn).toBe(true)
  component.find('button').simulate('click')
  expect(instance.state.isPlayer1Turn).toBe(false)
})

test('should compute round after both players select their moves', () => {
  const playRoundMock = jest.fn()
  const component = shallow(
    <Round
      player1="p1"
      player2="p2"
      rounds={[]}
      moves={DEFAULT_RULES.moves}
      handlePlayRound={playRoundMock}
    />
  )

  component.find('button').simulate('click')
  component.find('button').simulate('click')

  expect(playRoundMock).toHaveBeenCalled()
})
