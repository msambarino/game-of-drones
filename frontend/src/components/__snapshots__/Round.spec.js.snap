// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  className="container"
>
  <section
    className="section is-small columns"
  >
    <div
      className="column"
    >
      <h1
        className="title"
      >
        Round 1
      </h1>
      <h3
        className="subtitle"
      >
        p1's Turn
      </h3>
      <div
        className="field is-horizontal"
      >
        <label
          className="label field-label is-normal"
        >
          Select Move:
        </label>
        <div
          className="field-body"
        >
          <div
            className="field is-narrow"
          >
            <div
              className="control"
            >
              <div
                className="select is-fullwidth"
              >
                <select
                  onChange={[Function]}
                  value="rock"
                >
                  <option
                    key="rock"
                    value="rock"
                  >
                    rock
                  </option>
                  <option
                    key="paper"
                    value="paper"
                  >
                    paper
                  </option>
                  <option
                    key="scissors"
                    value="scissors"
                  >
                    scissors
                  </option>
                </select>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="control"
      >
        <button
          className="button is-primary"
          onClick={[Function]}
        >
          Ok
        </button>
      </div>
    </div>
    <div
      className="column"
    >
      <h3
        className="subtitle"
      >
        Score
      </h3>
      <Score
        player1="p1"
        player2="p2"
        rounds={Array []}
      />
    </div>
  </section>
</div>
`;
